### PR TITLE
Fix optional list error when class or value is none

### PR DESCRIPTION
**Pull request by tjeerddie**
Changed so the `ValueError` that shows `'NoneType' object has no attribute 'name'` when you have a list that does not have `None` to a `ValidationError`.

- add tests for direct relations in `api/test_subsctiptions`

### CODE REVIEW

File: orchestrator/domain/base.py [hunk 1]
```diff
@@ -387,13 +387,15 @@ def _data_from_lifecycle(cls, other: "DomainModel", status: SubscriptionLifecycl
                         possible_product_block_types[item.name]._from_other_lifecycle(item, status, subscription_id)
                     )
             else:
-                if is_optional_type(field_type):
+                if is_optional_type(field_type) and not is_union_type(field_type):
```
Comment (tjeerddie) [inline orchestrator/domain/base.py:390]: This might be unnecessary

Comment (pboers1988) [reply orchestrator/domain/base.py:390]: No Looks good I think,

File: orchestrator/domain/base.py [hunk 2]
```diff
@@ -387,13 +387,15 @@ def _data_from_lifecycle(cls, other: "DomainModel", status: SubscriptionLifecycl
...
                     data[field_name] = None
                     field_type = first(get_args(field_type))
                 elif is_union_type(field_type):
+                    if is_optional_type(field_type):
+                        data[field_name] = None
                     field_types = get_args(field_type)
                     for f_type in field_types:
-                        if f_type.name == value.name:
+                        if value and not isinstance(None, f_type) and f_type.name == value.name:
```
Comment (Mark90) [inline orchestrator/domain/base.py:398]: `not isinstance(None, f_type)` can None be subclassed?

Comment (tjeerddie) [reply orchestrator/domain/base.py:398]: `f_type` can be a `NoneType`, don't know a better way to validate that

Comment (Mark90) [reply orchestrator/domain/base.py:398]: And `if value` is not ideal for falsy values

Comment (Mark90) [reply orchestrator/domain/base.py:398]: > `f_type` can be a `NoneType` class, don't know a better way to validate that

Ah, I see that `types.NoneType` was added in 3.10. Yeah, makes sense then :)

**Review by Mark90 — APPROVED** [review]

**Pull request merged**